README.md
# animalsgonewild

[![codecov](https://codecov.io/gh/UBC-MDS/animalsgonewild/branch/main/graph/badge.svg?token=tGLNiVr2OZ)](https://codecov.io/gh/UBC-MDS/animalsgonewild) [![ci-cd](https://github.com/UBC-MDS/animalsgonewild/actions/workflows/ci-cd.yml/badge.svg)](https://github.com/UBC-MDS/animalsgonewild/actions/workflows/ci-cd.yml) [![Docs](https://readthedocs.org/projects/animalsgonewild/badge/?version=latest)](https://animalsgonewild.readthedocs.io/en/latest/index.html)

This package is designed to demonstrate how basic features of text analysis can be utilized to analyze and represent a text file or string while applying a humorous lens (because what is data science without a dash of humor!). It counts the number of words from a text, calculates the average word length of that text, and returns an animal image corresponding to the average word length. It subsequently transforms the text into a wordcloud in the shape of the animal. We have included a bonus fourth function for fun, inspired by the popular childhood game - MadLibs.

# Contributors

Kyle Maj, Nagraj Rao, Morgan Rosenberg, Junrong Zhu

## Installation

```bash
pip install animalsgonewild
```

## Usage

This package can be used in conjunction with any code to read in multiple text files to analyze and compare whole corpuses (your software must call the Animals Gone Wild functions for each text variable).

For more examples of usage in python, please check out this [link](https://animalsgonewild.readthedocs.io/en/latest/example.html).

```
import animalsgonewild
from animalsgonewild.animalsgonewild import animalClassifier
from animalsgonewild.animalsgonewild import animalType
from animalsgonewild.animalsgonewild import textTransformer
from animalsgonewild.animalsgonewild import wordClous
```

### Function 1: animalClassifier

This function takes a sequence of text(str), counts the words in the string, and then returns an animal type (str).

### Function 2: animalType

This function takes a sequence of text(str)  and a species (str), determines the average word length (proxy for language complexity), and returns an smart looking animal image (jpg) corresponding to the average word length.

### Function 3: wordcloud

This function takes a sequence of text(str) and an animal image (jpg), and returns a wordcloud in the shape of the species comprised of the sequence of text (jpg).

### Function 4: textTransformer

This function takes a sequence of text(str) and a species(str), and returns a new text sequence where all proper nouns are replaced with the species.

### Fit within the Python ecosystem

While other fun packages with animal images exist, most are very basic. For example, the animal-cuties script (<https://pypi.org/project/animal-cuties/#description>) generates animal images (e.g. animal-cuties cat). However, we were unable to find any interactive, multidimensional comedic relief package, where users can input information, and receive dynamic humorous feedback in the form of cute and/or hilarious animals. Given how much time we all are spending in front of a computer screen during the pandemic, this package is an essential addition to the ecosystem promoting mental wellness through humor. By offering it as a package rather than a script, we also empower other developers to integrate this is as a fun injection to their coding projecs.

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`animalsgonewild` was created by DSCI 524 - Team 16:
Nagraj Rao, Junrong Zhu, Kyle Maj, Morgan Rosenberg

It is licensed under the terms of the MIT license.

## Credits

`animalsgonewild` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).

# Easter egg #1

Did you know the name of our package was almost called "whattheduck"?

# Easter egg #2

Got any carrots?

# Easter egg #3

Aren't we pretty?


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.1.8 (2022-02-05)


## v1.1.7 (2022-02-05)


## v1.1.6 (2022-02-05)


## v1.1.5 (2022-02-05)


## v1.1.4 (2022-02-04)


## v1.1.3 (2022-02-04)


## v1.1.2 (2022-02-03)


## v1.1.0 (29/01/2022)

### Feature

- implemented ReadtheDocs documentation
- configured ci-cd.yml for Github actions

### Fix

- corrected a file pathing issue which was causing tests to fail

## v1.0.0 (22/01/2022)

### Feature

- Added functionality to the animalsgonewild functions

### Fix

- Fixed an issue with file pathing within the functions

### Tests

- Added new tests to all package modules in test folder

## v0.1.0 (11/01/2022)

- First release of `animalsgonewild`!

tests/test_wordCloud.py
from animalsgonewild.animalsgonewild import wordCloud
import pytest

text_wordcloud = "https://www.canada.ca/en.html"
text_wordcloud1 = "http://www.calendar.ubc.ca/"
text_wordcloud2 = "https://en.wikipedia.org/wiki/Outside_(magazine)"
text_wordcloud3 = "https://www.britannica.com/place/Japan"

def test_duckshapedwordcloud():
    expected = 'duck'
    actual = wordCloud(text_wordcloud)
    assert actual == expected, "WordCloud failed to output duck!"

def test_monkeyshapedwordcloud():
    expected = 'monkey'
    actual = wordCloud(text_wordcloud1)
    assert actual == expected, "WordCloud failed to output monkey!"

def test_giraffeshapedwordcloud():
    expected = 'giraffe'
    actual = wordCloud(text_wordcloud2)
    assert actual == expected, "WordCloud failed to output giraffe!"

def test_whaleshapedwordcloud():
    expected = 'whale'
    actual = wordCloud(text_wordcloud3)
    assert actual == expected, "WordCloud failed to output whale!"

def test_input_error_wordcloud():
    """Check TypeError raised when the input is not string"""
    with pytest.raises(TypeError):
        number = 111
        wordCloud(number)

tests/test_animalType.py
from animalsgonewild.animalsgonewild import animalType
from pathlib import Path
from PIL import Image
import pytest

big_words = "inscrutable or at least esoteric language."
small_words = "hi my name is john and I am simple."
formatted_text = open(Path("tests/bigwords.txt")).read()
try_num = 111

def test_big_words():
    expected = Image.open("tests/smart_Duck.jpeg")
    actual = animalType("Duck", big_words)
    assert expected == actual, "Failed to return smart Duck"

def test_small_words():
    expected = Image.open("tests/dumb_Giraffe.jpeg")
    actual = animalType("Giraffe", small_words)
    assert expected == actual, "Failed to return dumb Giraffe"
        
def test_file_text():
    expected = Image.open("tests/smart_Whale.jpeg")
    actual = animalType("Whale", formatted_text)
    assert expected == actual, "Failed to return smart Whale"
    
def test_input_error():
    """Check TypeError raised when the input is not string"""
    with pytest.raises(TypeError):
        animalType("Duck", try_num)
    with pytest.raises(TypeError):
        animalType(try_num, "this is a string")

tests/test_animalsgonewild.py
#from animalsgonewild import animalsgonewild


src/wordcloud_new.py
from newspaper import Article
from wordcloud import STOPWORDS
from wordcloud import WordCloud
import matplotlib.pyplot as plt
from PIL import Image
import numpy as np

def wordCloud(link):
    """
    Inputs URL of a website to return the name and a wordcloud in the shape of the animal created using the text.

    Parameters
    ----------
    link: str
        URL of the webpage from which the animal shaped wordcloud is to be created

    Returns
    -------
    jpg
        a wordcloud image in the shape of the animal based on number of words in the article obtained from the URL.
    txt
        name of the animal associated with the jpg output.

    Example
    --------
    >>> wordCloud('https://www.britannica.com/place/Japan') 
    whale
    ![image]('animalsgonewild/images/whale.jpg')
    """
    if not isinstance(link, str):
        raise TypeError("Text input should be of type 'str' and a URL")

    article = Article(link)
    article.download()
    article.parse()
    article_length = len(article.text)
    
    if article_length < 300:
        mask = np.array(Image.open('animalsgonewild/images/duck.jpg'))
        wc = WordCloud(stopwords=STOPWORDS, mask=mask, background_color="white", max_words=2000, max_font_size=256, random_state=42, width=mask.shape[0], height=mask.shape[0])
        wc.generate(article.text)
        plt.imshow(wc, interpolation="bilinear")
        plt.axis('off')
        animal = 'duck'
    elif article_length < 2000:
        mask = np.array(Image.open('animalsgonewild/images/monkey.jpg'))
        wc = WordCloud(stopwords=STOPWORDS, mask=mask, background_color="white", max_words=2000, max_font_size=256, random_state=42, width=mask.shape[0], height=mask.shape[0])
        wc.generate(article.text)
        plt.imshow(wc, interpolation="bilinear")
        plt.axis('off')
        animal = 'monkey'
    elif article_length < 3000:
        mask = np.array(Image.open('animalsgonewild/images/giraffe.jpg'))
        wc = WordCloud(stopwords=STOPWORDS, mask=mask, background_color="white", max_words=2000, max_font_size=256, random_state=42, width=mask.shape[0], height=mask.shape[0])
        wc.generate(article.text)
        plt.imshow(wc, interpolation="bilinear")
        plt.axis('off')
        animal = 'giraffe'
    else:
        mask = np.array(Image.open('animalsgonewild/images/whale.jpg'))
        wc = WordCloud(stopwords=STOPWORDS, mask=mask, background_color="white", max_words=2000, max_font_size=256, random_state=42, width=mask.shape[0], height=mask.shape[0])
        wc.generate(article.text)
        plt.imshow(wc, interpolation="bilinear")
        plt.axis('off')
        animal = 'whale'
    return animal

src/textTransformer.py
import nltk
nltk.download('wordnet')
nltk.download('omw-1.4')
from nltk.corpus import wordnet as wn
import string


def textTransformer(text):
    """    
    This function modifies a block of text by replacing all nouns
    with an animal.

    Parameters
    ----------
    text: str
        text to be transformed.

   Returns
    -------
    str
        a string containing the transformed text.

    Example
    --------
    >>> text = "The silly chicken walked across the road"
    >>> textTransformer(text)
    "The silly duck walked across the duck"
    """

    #animal = animalClassifier(text)   # use this once animalClassfier is working

    animal = "Duck"   #temporary paceholder

    text_list = text.split()
    revised_text_list = text_list

    for i in range(0, len(text_list)):
        if text_list[i] in nouns:
            revised_text_list[i] = animal

    output = ' '.join(revised_text_list)
        
    return output

src/animalClassifier.py
from pickle import TRUE
import string


def animalClassifier(text):
    """
    Count total unique words in the input text and returns an animal type.
    
    Parameters
    ----------
    text: str
        text to be analyzed

    Returns
    -------
    str
        the corresponding animal species
        
    Example
    --------
    >>> animalClassifier("what do you think?")
    "duck"
    """
    if not isinstance(text, str):
        raise TypeError("Text input should be of type 'str'")
    
    animal_list = ["Duck", "Monkey", "Giraffe", "Whale"]
    cleaned_text = text.translate(str.maketrans('', '', string.punctuation))
    total_words = len(str.split(cleaned_text))
    unique_words = np.unique(str.split(cleaned_text))
    uni_counts = len(unique_words)


    if uni_counts/total_words <= 0.4:
        output = animal_list[0]
    elif uni_counts/total_words <= 0.6:
        output = animal_list[1]
    elif uni_counts/total_words <= 0.8:
        output = animal_list[2]
    else:
        output = animal_list[3]

    return output

src/animalsgonewild/animalsgonewild.py
import string
import numpy as np
import nltk
nltk.download('wordnet')
nltk.download('omw-1.4')
from pathlib import Path
from PIL import Image
from nltk.corpus import wordnet as wn
from newspaper import Article
from wordcloud import STOPWORDS
from wordcloud import WordCloud
import matplotlib.pyplot as plt

def animalClassifier(text):
    """
    Count total unique words in the input text and returns an animal type.
    
    Parameters
    ----------
    text: str
        text to be analyzed

    Returns
    -------
    str
        the corresponding animal species
        
    Example
    --------
    >>> animalClassifier("what do you think?")
    "duck"
    """
    if not isinstance(text, str):
        raise TypeError("Text input should be of type 'str'")
    
    animal_list = ["Duck", "Monkey", "Giraffe", "Whale"]
    cleaned_text = text.translate(str.maketrans('', '', string.punctuation))
    total_words = len(str.split(cleaned_text))
    unique_words = np.unique(str.split(cleaned_text))
    uni_counts = len(unique_words)


    if uni_counts/total_words <= 0.4:
        output = animal_list[0]
    elif uni_counts/total_words <= 0.6:
        output = animal_list[1]
    elif uni_counts/total_words <= 0.8:
        output = animal_list[2]
    else:
        output = animal_list[3]

    return output
    
def textTransformer(text):
    """    
    This function modifies a block of text by replacing all nouns with an animal.

    Parameters
    ----------
    text: str
        text to be transformed.

   Returns
    -------
    str
        a string containing the transformed text.

    Example
    --------
    >>> text = "The silly chicken walked across the road"
    >>> textTransformer(text)
    "The smart duck walked across the duck"
    """
    
    if not isinstance(text, str):
        raise TypeError("Text input should be of type 'str'")
    
    nouns = {x.name().split('.', 1)[0] for x in wn.all_synsets('n')}
    animal = animalClassifier(text)

    text_list = text.split()
    revised_text_list = text_list

    for i in range(0, len(text_list)):
        if text_list[i] in nouns:
            revised_text_list[i] = animal

    output = ' '.join(revised_text_list)
    
    assert type(text) == str, "Output is not str datatype"
    return output



def animalType(species, text):
    """    
    Analyzes text based on average word length and returns an animal picture representing language complexity. 
    
    Parameters
    ----------
    species: str
        name of the species determined from animalClassifier.py or user determined
    text: str
        text to be analyzed
    Returns
    -------
    jpeg
        an image representative of the species and the complexity of the words used (more complex words will have 'wiser' looking animals)
    
    Example
    --------
    >>> animal_rep = animalType(animalClassifier(text), text) 
    >>> animal_rep = animalType("duck", text) 
    >>> animal_rep(species, text)
    smart_Duck.jpeg
    """
    
    if not isinstance(text, str):
        raise TypeError("Text input should be of type 'str'")
        
    if not isinstance(species, str):
        raise TypeError("Species input should be of type 'str'")
    
    words = text.translate(str.maketrans('', '', string.punctuation)).split()
    average = sum(len(w) for w in words) / len(words)
    iq = ""
    
    if average < 5:
        iq = "dumb"
    else:
        iq = "smart"
    filename = "".join([iq, "_", species, ".jpeg"])    
    
    data_folder = Path("src/animalsgonewild/imgs/")

    file_to_open = data_folder / filename

    img = Image.open(file_to_open)

    return img


def wordCloud(link):
    """
    Inputs the URL of a website to return a wordcloud in the shape of the animal created and the animal name itself.
    Parameters
    ----------
    text: str
        text to be inserted
    Returns
    -------
    jpg
        a wordcloud image in the shape of the species inputed.
    Example
    --------
    >>> word_cloud = wordCloud('https://www.britannica.com/place/Japan') 
    whale
    """
    if not isinstance(link, str):
        raise TypeError("Text input should be of type 'str' and a URL")
    article = Article(link)
    article.download()
    article.parse()
    article_length = len(article.text)
    
    if article_length < 300:
        mask = np.array(Image.open(Path('src/animalsgonewild/images/duck.jpg')))
        wc = WordCloud(stopwords=STOPWORDS, mask=mask, background_color="white", max_words=2000, max_font_size=256, random_state=42,
                       width=mask.shape[0], height=mask.shape[0])
        wc.generate(article.text)
        plt.imshow(wc, interpolation="bilinear")
        plt.axis('off')
        animal = 'duck'
    elif article_length < 2000:
        mask = np.array(Image.open(Path('src/animalsgonewild/images/monkey.jpg')))
        wc = WordCloud(stopwords=STOPWORDS, mask=mask, background_color="white", max_words=2000, max_font_size=256, random_state=42,
                       width=mask.shape[0], height=mask.shape[0])
        wc.generate(article.text)
        plt.imshow(wc, interpolation="bilinear")
        plt.axis('off')
        animal = 'monkey'
    elif article_length < 3000:
        mask = np.array(Image.open(Path('src/animalsgonewild/images/giraffe.jpg')))
        wc = WordCloud(stopwords=STOPWORDS, mask=mask, background_color="white", max_words=2000, max_font_size=256, random_state=42,
                       width=mask.shape[0], height=mask.shape[0])
        wc.generate(article.text)
        plt.imshow(wc, interpolation="bilinear")
        plt.axis('off')
        animal = 'giraffe'
    else:
        mask = np.array(Image.open(Path('src/animalsgonewild/images/whale.jpg')))
        wc = WordCloud(stopwords=STOPWORDS, mask=mask, background_color="white", max_words=2000, max_font_size=256, random_state=42,
                       width=mask.shape[0], height=mask.shape[0])
        wc.generate(article.text)
        plt.imshow(wc, interpolation="bilinear")
        plt.axis('off')
        animal = 'whale'
    return animal
